### PR TITLE
Adding proper support for fid_nic

### PR DESCRIPTION
**Pull request by shefty**

### CODE REVIEW

**Comment by jsquyres** [comment]
So just to be clear: this PR doesn't (yet?) add a field for the fid_nic, right?

**Comment by shefty** [comment]
Correct -- I have a local patch in progress for that.  I need to add compat code still

**Comment by shefty** [comment]
I've updated this PR with a commit that adds fid_nic directly to fi_info, along with the proper compatibility code (I hope).

There is still some work to do.  The fi_getinfo/fi_freeinfo/fi_tostr/fi_dupinfo calls need to be updated to process the fi_info::nic field.

File: src/fi_tostr.c [hunk 1]
```diff
@@ -537,6 +550,114 @@ static void ofi_tostr_fabric_attr(char *buf, const struct fi_fabric_attr *attr,
...
+	ofi_strncatf(buf, len, "%saddress: %s\n", prefix, attr->address);
+	ofi_strncatf(buf, len, "%smtu: %zu\n", prefix, attr->mtu);
+	ofi_strncatf(buf, len, "%sspeed: %zu\n", prefix, attr->speed);
+	ofi_strncatf(buf, len, "%sstate: ", prefix);
+	ofi_tostr_link_state(buf, len, attr->state);
+	ofi_strncatf(buf, len, "\n%snetwork_type: %s\n", prefix,
+		     attr->network_type);
+}
+
+/* TODO: ofi_strcatf assumes that we're concatenation to the end of
```
Comment (bturrubiates) [inline src/fi_tostr.c]: This comment looks like it could be removed after the addition of `ofi_strncatf`.

Comment (shefty) [reply src/fi_tostr.c]: right - I'll remove

File: src/fi_tostr.c [hunk 2]
```diff
@@ -76,6 +76,10 @@
 	case SYM: { ofi_strcatf(buf, #SYM); break; }
 #define IFFLAGSTR(flags, SYM) \
 	do { if (flags & SYM) ofi_strcatf(buf, #SYM ", "); } while(0)
+#define CASEENUMSTRN(SYM, N) \
+	case SYM: { ofi_strncatf(buf, N, #SYM); break; }
+#define IFFLAGSTRN(flags, SYM, N) \
+	do { if (flags & SYM) ofi_strcatf(buf, N, #SYM ", "); } while(0)
```
Comment (shefty) [inline src/fi_tostr.c]: This should have called ofi_strncatf().  It's been updated (not pushed yet).

**Comment by shefty** [comment]
Updated -- fixed crash dealing with nested va_args.  Added fid_nic handling to dupinfo/freeinfo/tostr.

I think this is ready as a complete solution now.  It just needs a provider to use it.

**Pull request merged**